kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/fifo/Fifo.java

```diff
@@ -0,0 +1,180 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Eurotech and/or its affiliates and others
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *  Eurotech
+ *
+ *******************************************************************************/
+package org.eclipse.kura.internal.wire.fifo;
+
+import static java.util.Objects.isNull;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.kura.configuration.ConfigurableComponent;
+import org.eclipse.kura.configuration.ConfigurationService;
+import org.eclipse.kura.localization.LocalizationAdapter;
+import org.eclipse.kura.localization.resources.WireMessages;
+import org.eclipse.kura.wire.WireEmitter;
+import org.eclipse.kura.wire.WireEnvelope;
+import org.eclipse.kura.wire.WireHelperService;
+import org.eclipse.kura.wire.WireReceiver;
+import org.eclipse.kura.wire.WireSupport;
+import org.osgi.service.wireadmin.Wire;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Fifo implements WireEmitter, WireReceiver, ConfigurableComponent {
+
+    private static final String DISCARD_ENVELOPES_PROP_NAME = "discard.envelopes";
+    private static final String QUEUE_CAPACITY_PROP_NAME = "queue.capacity";
+
+    private static final Logger logger = LoggerFactory.getLogger(Fifo.class);
+    private static final WireMessages message = LocalizationAdapter.adapt(WireMessages.class);
+
+    private volatile WireHelperService wireHelperService;
+    private WireSupport wireSupport;
+
+    private FifoEmitterThread emitterThread;
+
+    public void bindWireHelperService(final WireHelperService wireHelperService) {
+        if (isNull(this.wireHelperService)) {
+            this.wireHelperService = wireHelperService;
+        }
+    }
+
+    public void unbindWireHelperService(final WireHelperService wireHelperService) {
+        if (this.wireHelperService == wireHelperService) {
+            this.wireHelperService = null;
+        }
+    }
+
+    public void activate(final Map<String, Object> properties) {
+        logger.info(message.activatingFifo());
+        wireSupport = this.wireHelperService.newWireSupport(this);
+        updated(properties);
+        logger.info(message.activatingFifoDone());
+    }
+
+    public void deactivate() {
+        logger.info(message.deactivatingFifo());
+        stopEmitterThread();
```

Comment:
We discussed about this and for now we decided to neither interrupt nor kill the emitter thread, because otherwise there is the risk of leaving downstream components in a bad state.
The current implementation notifies the emitter thread to stop using the FifoEmitterThead.shutdown() method, if this method is called the emitter dispatches at most one more envelope and then exits. Any threads blocked trying to add a new envelope to the queue will be notified of the fact that queue is being shut down and will unblock.
It is possible that the emitter thread never exits due to problems in the downstream components, but it's hard to recover in a clean way from such situation. Interrupting or killing the thread can cause trouble because we don't know what the components are doing.
Anyway we can change this in future if needed.